streamz/graph.py

```diff
@@ -21,107 +21,115 @@ def _clean_text(text, match=None):
     return text
 
 
-def create_graph(node, graph, prior_node=None, pc=None):
-    """Create graph from a single node, searching up and down the chain
+def build_node_set(node, s=None):
+    """Build a set of all the nodes in a streamz graph
 
     Parameters
     ----------
-    node: Stream instance
-    graph: networkx.DiGraph instance
+    node : Stream
+        The node to use as a starting point for building the set
+    s : set or None
+        The set to put the nodes into. If None return a new set full of nodes
+
+    Returns
+    -------
+    s : set
+        The set of nodes in the graph
+
     """
-    if node is None:
+    if s is None:
+        s = set()
+    if node is None or (
+        node in s
+        and all(n in s for n in node.upstreams)
+        and all(n in s for n in node.downstreams)
+    ):
         return
-    t = hash(node)
-    graph.add_node(t,
-                   label=_clean_text(str(node)),
-                   shape=node._graphviz_shape,
-                   orientation=str(node._graphviz_orientation),
-                   style=node._graphviz_style,
-                   fillcolor=node._graphviz_fillcolor)
-    if prior_node:
-        tt = hash(prior_node)
-        if graph.has_edge(t, tt):
-            return
-        if pc == 'downstream':
-            graph.add_edge(tt, t)
-        else:
-            graph.add_edge(t, tt)
-
-    for nodes, pc in zip([list(node.downstreams), list(node.upstreams)],
-                         ['downstream', 'upstreams']):
-        for node2 in nodes:
-            if node2 is not None:
-                create_graph(node2, graph, node, pc=pc)
+    s.add(node)
+    s.update(node.upstreams)
+    s.update({n for n in node.downstreams})
+    [build_node_set(n, s) for n in list(s)]
```

Comment:
Shouldn't this be done only for *new* nodes in the set? I know that sets nicely dedup, but seems like we could have a lot unnecessary recursion here.